bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/handler/SMAEnergyMeterHandler.java

```diff
@@ -61,68 +74,121 @@ public void initialize() {
 
         EnergyMeterConfig config = getConfigAs(EnergyMeterConfig.class);
 
-        int port = (config.getPort() == null) ? EnergyMeter.DEFAULT_MCAST_PORT : config.getPort();
-        energyMeter = new EnergyMeter(config.getMcastGroup(), port);
+        int port = (config.getPort() == null) ? PacketListener.DEFAULT_MCAST_PORT : config.getPort();
+
         try {
-            energyMeter.update();
+            PacketListener thingListener = listenerRegistry.getListener(config.getMcastGroup(), port);
+            serialNumber = config.getSerialNumber();
+            if (serialNumber == null) {
+                if (!thing.getProperties().containsKey(Thing.PROPERTY_SERIAL_NUMBER)
+                        || (serialNumber = retrieveSerialNumber(thing.getProperties())) == null) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
+                            "Meter serial number missing");
+                    return;
+                }
+
+                // copy serial number from thing properties into config
+                Map<String, Object> newConfig = editConfiguration().getProperties();
+                newConfig.put("serialNumber", serialNumber);
+                Thing thing = editThing().withConfiguration(new Configuration(newConfig)).build();
+                updateThing(thing);
+                updateStatus(ThingStatus.OFFLINE);
+                return;
+            } else {
+                if (!thing.getProperties().containsKey(Thing.PROPERTY_SERIAL_NUMBER)) {
+                    Map<String, String> props = editProperties();
+                    props.put(Thing.PROPERTY_VENDOR, "SMA");
+                    props.put(Thing.PROPERTY_SERIAL_NUMBER, serialNumber);
+                    updateProperties(props);
+                }
+            }
 
-            updateProperty(Thing.PROPERTY_VENDOR, "SMA");
-            updateProperty(Thing.PROPERTY_SERIAL_NUMBER, energyMeter.getSerialNumber());
-            logger.debug("Found a SMA Energy Meter with S/N '{}'", energyMeter.getSerialNumber());
+            logger.debug("Activated handler for SMA Energy Meter with S/N '{}'", serialNumber);
+
+            this.updateInterval = TimeUnit.SECONDS
+                    .toMillis(config.getPollingPeriod() == null ? 30 : config.getPollingPeriod());
+            timeoutFuture = scheduler.scheduleWithFixedDelay(this, updateInterval * 2, updateInterval,
+                    TimeUnit.MILLISECONDS);
```

Comment:
Rather than calling this with you class extending Runnable, just call your method `run` there and no need to implement Runnable.

Reply:
For clarity of stack trace I prefer implementing interfaces and not having lambdas or anonymous clases with generated names.